File: Client_WPF/BetLocation.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Text.Json.Serialization;
using System.Threading.Tasks;

namespace Client_WPF
{
    public class BetLocation
    {
        private string location;
        private Horse[] horses = new Horse[5];
        private float trackLength;
        private bool gameFinished;
        private Horse winner;
        private bool gameStarted;

        public BetLocation() { }

        public BetLocation(string location, Horse[] horses, float trackLength, bool gameFinsihed, Horse winner, bool gameStarted)
        {
            this.location = location;
            this.horses = horses;
            this.trackLength = trackLength;
            this.gameFinished = gameFinsihed;
            this.winner = winner;
            this.gameStarted = gameStarted;
        }

        //TODO: REFACTOR SO THAT IT WILL EVENTUALLY WORK

        public async Task<Dictionary<string, float>> CalculateBetResults()
        {
            Dictionary<string, float> betResults = new Dictionary<string, float>();

            // Find the winning horse
            Horse? winningHorse = horses.FirstOrDefault(horse => horse.RunDistance > trackLength);

            // Iterate through each horse's bets
            foreach (Horse horse in horses)
            {
                foreach (KeyValuePair<string, float> bet in horse.Bets)
                {
                    // Check if the bettor's horse won
                    bool betWon = horse == winningHorse;

                    // Calculate the result based on the bet
                    float result = betWon ? bet.Value : -bet.Value;

                    // Update the bet result for the bettor
                    if (betResults.ContainsKey(bet.Key))
                    {
                        betResults[bet.Key] += result;
                    }
                    else
                    {
                        betResults[bet.Key] = result;
                    }
                }
            }

            return betResults;
        }




        public override string ToString()
        {
            return $"LOCATION: {location} horses: {horses.Length} tracklength: {trackLength}";
        }


        [JsonPropertyName("location")]
        public string Location
        {
            get { return location; }
            set { location = value; }
        }

        [JsonPropertyName("horses")]
        public Horse[] Horses
        {
            get { return horses; }
            set { horses = value; }
        }

        [JsonPropertyName("trackLength")]
        public float TrackLength
        {
            get { return trackLength; }
            set { trackLength = value; }
        }

        [JsonPropertyName("gameFinished")]
        public bool GameFinished
        {
            get { return gameFinished; }
            set { gameFinished = value; }
        }

        [JsonPropertyName("winner")]
        public Horse Winner
        {
            get { return winner; }
            set { winner = value; }
        }

        [JsonPropertyName("gameStarted")]
        public bool GameStarted
        {
            get { return gameStarted; }
            set { gameStarted = value; }
        }
    }
}

File: webclient_maier/src/Utitlity/BetLocation.ts
import User from "./User";
import Horse from './Horse';

export default class BetLocation{

  private _location: string;
  private _horses: Horse[];
  private _trackLength:number;
  private _gameFinished: boolean;
  private _winner: Horse;
  private _gameStarted: boolean;

    constructor(location: string, horses:Horse[], trackLength:number, gameFinished: boolean, winner:Horse, gameStarted:boolean){
        this._trackLength = trackLength;
        this._location = location;
        this._horses = horses;
        this._gameFinished = gameStarted;
        this._winner = winner;
        this._gameStarted = gameStarted;
    }

    public async GetWinner(): Promise<Record<string, number>>{
      let winners : Record<string, number> = {};

      this._horses.forEach((winningHorse: Horse) => {
          Object.entries(winningHorse.bets).forEach(([key, value]) => {

              // Check if the bettor's horse won
              let betWon: boolean = this._winner == winningHorse;

              // Calculate the result based on the bet
              let result: number = betWon ? value : -value;

              // Update the bet result for the bettor
              if (key in winners)
              {
                winners[key] += result;
              }
              else
              {
                console.log(result);
                winners[key] = result;
              }
            }
          );
        }
      );


      return winners;
    }



    public get location():string{
      return this._location;
    }

    public set location(location: string){
      this._location = location;
    }

    public get horses(): Horse[] {
      return this._horses;
    }

    public set horses(horses:Horse[]){
      this._horses = horses;
    }

    public get trackLength(){
      return this._trackLength;
    }

    public set trackLength(trackLength:number){
      this._trackLength = trackLength;
    }

    public get gameFinished(){
      return this._gameFinished;
    }

    public set gameFinished(gameFinished:boolean){
      this._gameFinished = gameFinished;
    }

    public get winner(){
      return this._winner
    }

    public set winner(winner:Horse){
      this._winner = winner;
    }

    public get gameStarted(){
      return this._gameStarted;
    }

    public set gameStarted(gameStarted:boolean){
      this._gameStarted = gameStarted;
    }

    public toString(): void{
      console.log("location: " + this.location + " horsesize: " + this.horses.length + " gamefinsihed: " + this.gameFinished + " gameStarted: " + this.gameStarted + " winner: " + this.winner);
}

}
